Symbol tags in SymbolInformation, WorkspaceSymbol, CallHierarchyItem and TypeHierarchyItem

ratzdi opened the pull request:
This PR is continuation of #167536.

What is new?

It adds code to compute symbol tags for the following types:
    [SymbolInformation](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#symbolInformation)
    [WorkspaceSymbol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceSymbol)
    [CallHierarchyItem](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#callHierarchyItem)
    [TypeHierarchyItem](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#typeHierarchyItem)

Calculation of the symbol tags `Overrides` and `Implements`
Calculation of symbol tags during the AST creation and embedding them in `Symbol`.

Minor change on `struct Symbol` to minimize paddings and thus reducing momory consumption.

Related issues:
- https://github.com/llvm/llvm-project/pull/113669#issuecomment-2508048524
- https://github.com/llvm/llvm-project/issues/180843

Commit:
Fill symbol tags for the response TypeHierarchyItem of prepareTypeHierarchy.

Commit:
Fill symbol tags into the object SymbolInformation, and on getting workspace symbols.

Commit:
Minor improvements.

Commit:
Set the detail field of HierarchyItem.

Commit:
Collect symbols tags from AST in the method incomingCalls.

Commit:
Collect symbols tags from AST in the method outgoingCalls.

Commit:
Collect symbols tags from AST in the method subtypes.

Commit:
Collect symbols tags from AST in the method supertypes.

Commit:
Extended unit-tests to check the occurrence of symbol tags
in call-hierarchy and type-hierarchy.

Commit:
Simplify lambda capture by explicitly moving `Item` and parameters.
Fix code format.

ratzdi committed:
Feat: functions calculating the tags Overrides and Implements.

Changed the usage of SymbolTag:
- Abstract = pure virtual
- Virtual = a virtual method implemented in that class, not implementing or overriding a method from base class
- Implements = implements pure virtual method from base class
- Override = overriding a virtual method implemented in base class
- Final = final method

ratzdi committed:
Feat: compute symbol tags during the AST indexation.

ratzdi committed:
Merge branch 'llvm:main' into users/ratzdi/symbol_tags_in_call_type_hierarchies_and_symbol_info

ratzdi committed:
Change: removed AST usage in call/type hierarchy

ratzdi committed:
Feat: serialize/de-serialize symbol tags.

- Bump index version to 21
- Updated sample index
- Added tests

ratzdi committed:
Review: code review adjustments.

- avoid returning empty vector
- return either a filled vector or nullpt meaning there are no parents
- removed obsolete function
- texts rewritten correctly
- tightening the check to < 32

HighCommander4 reviewed (commented):
I haven't had a chance to look at this in any amount of detail yet, but just a high-level comment for now:

By adding a new field to `Symbol`, we are increasing the size of the index in memory and on disk. It would be helpful to understand the magnitude of this increase, by measuring the size of the index for a sizeable codebase (llvm-project is suitable, or another C++ project you work on if it's comparable in size to llvm-project or larger) before and after this patch.

The size of the index on disk can be measured by checking the size of the `<project>/.cache/clangd/index` directory after background indexing has completed.

The size of the index in memory can be measured by invoking the "clangd: Show memory usage" command in vscode, and looking at the `background_index` entry in the opened "Clangd memory usage" tree view.

ratzdi committed:
Change: relocated Tags to decrease padding.

64-bit

The problematic areas are:

SymInfo (6 bytes) → followed by a 2-byte gap
Origin (2 bytes) → followed by a 2-byte gap before StringRef
Tags (4 bytes) → followed by a 4-byte gap before SmallVector
Flags at the end → 7 bytes end padding

32-bit

SymInfo → 2-byte gap
Origin → 2-byte gap
Flags at the end → 3 bytes end padding

This patch reduces paddings:
64-bit: 240 -> 232 (-8 Bytes, ~3.3%)
32-bit: 144 -> 140 (-4 Bytes, ~2.8%)

timon-ul commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -159,6 +172,45 @@ SymbolTags toSymbolTagBitmask(const SymbolTag ST) {
+      if (Overridden->isVirtual() && !Overridden->isPureVirtual())
+        return true;
+    }
+    return false;
```
Kinda redundant.

timon-ul reviewed (changes requested):
Hi @ratzdi I have finally gotten around to review your pullrequest. Overall I do not see any major problems, but I also am not yet confident I fully understand all of the implications these changes have.

That aside I see you have enabled Copilot Review, I think I have seen a discussion somewhere before that this is not allowed based on the [AI tool policy](https://llvm.org/docs/AIToolPolicy.html) ([...]automated review tools that publish comments without human review are not allowed."), so if you want to use AI to review your code please do so locally.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -208,21 +266,103 @@ SymbolTags computeSymbolTags(const NamedDecl &ND) {
 
+// Filter symbol tags based on the presence of other tags and the kind of
+// symbol. This is needed to avoid redundant tags.
+SymbolTags filterSymbolTags(const NamedDecl &ND, const SymbolTags ST) {
```
It is not always advisable to pass on the calculated symbol tags directly. For example, `void foo() final {}` would have, among others, the tags `virtual` and `overrides`. These two can be correctly reduced to a single symbol tag.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -159,6 +172,45 @@ SymbolTags toSymbolTagBitmask(const SymbolTag ST) {
   return (1 << static_cast<unsigned>(ST));
 }
 
+bool isOverrides(const NamedDecl *ND) {
```
done.

ratzdi commented on clang-tools-extra/clangd/index/SymbolCollector.cpp:
```diff
@@ -1126,6 +1127,9 @@ const Symbol *SymbolCollector::addDeclaration(const NamedDecl &ND, SymbolID ID,
     }
   };
+
+  S.Tags = computeSymbolTags(ND);
```
Collecting symbol tags is needed during processing headers and TUs.

ratzdi replied:
done.

ratzdi commented:
> > > preliminary results:
> > > clangd version > 	18.01.3 	this branch 	main @07ff1a63
> > > background index 	~1400 MB 	~2972 MB 	~2903 MB
> > > index cache disk usage 	~159 MB 	~188 MB 	~187 MB
> > 
> > 
> > Unrelated to this patch but, wow, that doubling of the index memory usage between 18.x and trunk is quite something. That probably deserves to be spun out into a separate issue and investigated further.
> 
> I cannot explain how the outlier occurred in the first measurement with 18.1.3. In the second measurement with 18.1.3, the memory/disc consumption is now more consistent.
> 
> current results:
>  clangd version > 	18.1.3 	this branch 	main @07ff1a63
> background index 	~2721 MB 	~2972 MB 	~2903 MB
> index cache disk usage 	~184 MB 	~188 MB 	~187 MB

Minor change on `struct Symbol` to minimize paddings:

64-bit

The problematic areas are:

SymInfo (6 bytes) → followed by a 2-byte gap
Origin (2 bytes) → followed by a 2-byte gap before StringRef
Tags (4 bytes) → followed by a 4-byte gap before SmallVector
Flags at the end → 7 bytes end padding

32-bit

SymInfo → 2-byte gap
Origin → 2-byte gap
Flags at the end → 3 bytes end padding


current results for the patch f99012b5:
 clangd version > | 18.1.3 | before f99012b5 | this from f99012b5 | main @07ff1a63
-- | -- | -- | -- | -- 
background index | ~2721 MB | ~2972 MB | ~2810 MB | ~2903 MB
index cache disk usage | ~184 MB | ~188 MB | ~188 MB | ~187 MB


This patch f99012b5 reduces paddings:
64-bit: 240 -> 232 (-8 Bytes, ~3.3%)
32-bit: 144 -> 140 (-4 Bytes, ~2.8%)

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -208,21 +320,47 @@ SymbolTags computeSymbolTags(const NamedDecl &ND) {
     return Tags;
 
+  // Apply specific filter to the symbol tags.
+  const auto filteredTags = filterSymbolTags(ND, symbolTags);
```
done.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -208,21 +320,47 @@ SymbolTags computeSymbolTags(const NamedDecl &ND) {
+}
+
 std::vector<SymbolTag> getSymbolTags(const NamedDecl &ND) {
   const auto symbolTags = computeSymbolTags(ND);
```
done.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -208,21 +320,47 @@ SymbolTags computeSymbolTags(const NamedDecl &ND) {
+  // Iterate through SymbolTag enum values and collect any that are present in
+  // the bitmask. SymbolTag values are in the numeric range
+  // [FirstTag .. LastTag].
+  constexpr unsigned MinTag = static_cast<unsigned>(SymbolTag::FirstTag);
```
refactored.

timon-ul replied:
Did this change to `uint32_t` get lost? (Also below this line there is another one which uses `unsigned`)

ratzdi replied:
done.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -208,21 +320,47 @@ SymbolTags computeSymbolTags(const NamedDecl &ND) {
   return Result;
 }
 
+std::vector<SymbolTag> expandTagBitmask(const SymbolTags STGS) {
```
Moved filterSymbolTags call into computeSymbolTags so that filtering is applied at the single point of tag computation.

timon-ul replied:
You are confusing me a bit, you moved the filtering into `getSymbolTags`, not into `computeSymbolTags`, which by the way is in my opinion what we want, the full Symbol information should still be in the index available. The filter is just for the client.

ratzdi replied:
> You are confusing me a bit, you moved the filtering into `getSymbolTags`, not into `computeSymbolTags`, which by the way is in my opinion what we want, the full Symbol information should still be in the index available. The filter is just for the client.

Sorry, just forgotten to update it after clarification.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -117,6 +120,16 @@ bool isAbstract(const Decl *D) {
+// virtual, even if it is not declared as such in the derived class.
+// Similarly, a method that implements a pure virtual method from a base class
+// is also considered virtual, even if it is not declared as such in the
+// derived class.
```
done.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -153,6 +202,63 @@ bool isUniqueDefinition(const NamedDecl *Decl) {
+    Result &= ~toSymbolTagBitmask(SymbolTag::Declaration);
+    Result &= ~toSymbolTagBitmask(SymbolTag::Definition);
+  }
+  if (ST & toSymbolTagBitmask(SymbolTag::Final)) {
```
Clarification is needed.

ratzdi replied:
Rolled back the wrong remove of Declaration and Definition from tags.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -153,6 +202,63 @@ bool isUniqueDefinition(const NamedDecl *Decl) {
+// Filter symbol tags based on the presence of other tags and the kind of
+// symbol. This is needed to avoid redundant tags, e.g. final implies override,
+// override implies virtual, etc.
+SymbolTags filterSymbolTags(const NamedDecl &ND, const SymbolTags ST) {
```
Clarification is needed.

timon-ul replied:
Okay let me desrcribe what I mean a bit more detailed:

For the first thing I was thinking about for example the case that we have a `final` method. So the bits for `virtual` `override` and `final` are set. Currently you check `override` before `final`, so since the `override` check is performed first we unset `virtual` and are done with the `if`. Then we reach the `final` check and now we unset the `override` bit, but also the already unset `virtual` bit. So essentially double the work which is not needed. So I think this could be done a bit more efficient.

For the second thing I was thinking about the fact that every single time you are calling this function you are calling `toSymbolTagBitmask(SymbolTag)` even though they all all constants, basically recomputing the constants all the time. That being said I am not sure if this really matters, the compiler might already optimize this out so it might be fine. I have no strong opinion on this right now.

ratzdi replied:
done.

ratzdi commented on clang-tools-extra/clangd/XRefs.cpp:
```diff
@@ -1872,6 +1873,7 @@ static std::optional<HierarchyItem> symbolToHierarchyItem(const Symbol &S,
   HI.name = std::string(S.Name);
   HI.detail = (S.Scope + S.Name).str();
   HI.kind = indexSymbolKindToSymbolKind(S.SymInfo);
+  HI.tags = expandTagBitmask(S.Tags);
```
The suggested change is suspicious, but the comment provides an important clue. The filtered tags are now stored in the index.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -208,21 +266,103 @@ SymbolTags computeSymbolTags(const NamedDecl &ND) {
+  if (symbolTags == 0)
+    return Tags;
+
+  // No filtering required since this function is only used for Symbols from the
+  // index, which have already been filtered in getSymbolTags(const NamedDecl
+  // &ND).
+
```
done.

ratzdi committed:
Review: code review adjustments.

Apply consistent SymbolTag filtering for indexed C++ methods

Introduce `getSymbolTags(const Symbol&)` and use it for index-backed symbol
paths so tag filtering is applied consistently outside AST-based code paths.

- add `getSymbolTags(const Symbol&)` in `FindSymbols` and document behavior
- apply method-specific tag filtering in:
  - workspace symbols (`FindSymbols.cpp`)
  - hierarchy items (`XRefs.cpp`)
- simplify `filterSymbolTags` to operate on `SymbolTags` only
- keep `Declaration`/`Definition` tags for virtual/override/implements cases
  (only remove tags that are semantically implied, e.g. `Virtual` by
  `Overrides`/`Implements`/`Abstract`)
- update unit tests in `FindSymbolsTests.cpp` and
  `SymbolCollectorTests.cpp` to assert the new expected tag sets
- use `getSymbolTags(...)` in tests instead of directly expanding raw bitmasks
- minor cleanup in `SymbolCollector::addDeclaration` tag assignment ordering

timon-ul commented on clang-tools-extra/clangd/index/Symbol.h:
```diff
@@ -22,6 +23,15 @@ namespace clangd {
+/// A bitmask type representing symbol tags supported by LSP.
+/// \see
+/// https://microsoft.github.io/language-server-protocol/specifications/specification-current/#symbolTag
+using SymbolTags = uint32_t;
```
I think it is fine to keep it like this. That being said, here my problem and thoughts about this setup:

Both types, `SymbolTag` and `SymbolTags` are conceptually **very** connected, but in the code, their connection is very loosely, only with the indirection of the `assert` below this. Preferrably to me if I change the type of `SymbolTag` the type of `SymbolTags` automatically adapts to it, so being a lot less fragile and involving manual changes to `SymbolTags` if we ever get there. Also it would be preferable if in `SymbolTags` you can only represent actual combinations of actual `SymbolTag`s. What I mean is that if I were to shift a bit 31 times it is a valid `uint32_t` and because of that a valid `SymbolTags`, but it does not represent any `SymbolTag`, because we do not have one for every bit. This again feels fragile.

Overall I think this is probably not that problematic, but just from a design perspective I am a bit unhappy about this. I leave it to you if you want to change this (or if a future reviewer wants this).

ratzdi replied:
done.

timon-ul commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -153,6 +192,64 @@ bool isUniqueDefinition(const NamedDecl *Decl) {
+  if (ST & toSymbolTagBitmask(SymbolTag::Overrides)) {
+    // Overrides means that ND overrides an existing implementation of a virtual
+    // method in a base class. If a symbol is marked as Overrides, the tags
+    // Virtual, Declaration and Definition should be removed, as the Overrides
```
Here and below you need to adjust your comments too! (Declaration & Definition are not removed anymore)

ratzdi replied:
done.

timon-ul commented on clang-tools-extra/clangd/index/SymbolCollector.cpp:
```diff
@@ -1138,6 +1139,7 @@ const Symbol *SymbolCollector::addDeclaration(const NamedDecl &ND, SymbolID ID,
   getSignature(*CCS, &Signature, &SnippetSuffix, SymbolCompletion.Kind,
                SymbolCompletion.CursorKind);
   S.Signature = Signature;
+  S.Tags = computeSymbolTags(ND);
```
You moved it further down which I think is not what you should be doing, in the case of us not wanting code completion we return early and thus before the symbol tags are computed, so for those cases we would have no symbol tags and that sounds incorrect to me.

ratzdi replied:
done.

ratzdi committed:
Review: code review adjustments.

Refine filterSymbolTags semantics for method-related tags

Improve SymbolTag normalization in `filterSymbolTags` by making tag
precedence explicit and handling redundant combinations consistently.

Changes:
- Introduce local bitmask constants (`Virtual`, `Overrides`, `Implements`,
  `Abstract`, `Final`) for readability and maintainability.
- Keep precedence logic explicit:
  - `Implements` removes `Overrides` and `Virtual`
  - `Overrides` removes `Virtual`
  - `Abstract` removes `Virtual`
  - `Final` removes redundant `Virtual`/`Overrides`
- Align inline comments with actual filtering behavior.

This makes tag filtering easier to reason about and avoids inconsistent
symbol-tag output for C++ methods.

ratzdi committed:
Refactor: Make SymbolTags type-safe and adapt tag handling

Replace raw integer-based SymbolTags with a dedicated strong type that is
tied to SymbolTag and sanitizes invalid bits. This removes a fragile coupling
between SymbolTag and SymbolTags and prevents representing non-existing tags.

Update clangd tag logic and serialization paths accordingly:
- switch tag construction/checks to SymbolTags::fromTag()/typed operations
- keep existing on-disk bit positions compatible
- adapt RIFF/YAML serialization to read/write raw SymbolTags storage safely
- update serialization tests to use the typed SymbolTags API

This improves robustness and maintainability while preserving behavior and
backward compatibility of serialized tag values.

ratzdi committed:
Fix: code format.

ratzdi committed:
Fix: code format.

timon-ul commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -242,11 +242,11 @@ template <typename E> constexpr E enumIncrement(E Value) {
 
 SymbolTags toSymbolTagBitmask(const SymbolTag ST) {
-  return (1 << static_cast<uint32_t>(ST));
+  return SymbolTags::fromTag(ST);
```
At this point, do we even still need this function?

ratzdi replied:
replaced.

timon-ul commented on clang-tools-extra/clangd/index/YAMLSerialization.cpp:
```diff
@@ -86,11 +86,28 @@ using clang::clangd::RelationKind;
 using clang::index::SymbolLanguage;
 using clang::tooling::CompileCommand;
 
+template <> struct ScalarTraits<SymbolTags> {
```
Where is this needed?

ratzdi replied:
In `clang-tools-extra/clangd/index/YAMLSerialization.cpp`  line 266 `IO.mapOptional("Tags", Sym.Tags)` to serialize/deserialize data to/from YAML files.

ratzdi replied:
change is obsolete, due to revert.

timon-ul commented on clang-tools-extra/clangd/index/YAMLSerialization.cpp:
```diff
@@ -86,11 +86,28 @@ using clang::clangd::RelationKind;
+
+  static StringRef input(StringRef Scalar, void *, SymbolTags &Value) {
+    SymbolTags::Storage Raw;
+    if (Scalar.getAsInteger(/*Radix=*/10, Raw))
```
Why a value of `10` ?

ratzdi replied:
The format remains unambiguous and stable (no 0x..., no octal, no prefix mixing). A round trip remains consistent: writing as a decimal number, reading as a decimal number. This prevents "random" interpretation of other notations if someone manually edits YAML.

ratzdi replied:
change is obsolete, due to revert.

timon-ul commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -194,43 +194,32 @@ bool isUniqueDefinition(const NamedDecl *Decl) {
+    ST &= ~VirtualMask;
+
+  // Final implies Virtual; Overrides is also redundant as Final overrides are still overrides.
+  if (ST & FinalMask)
```
Move this one above the `override` check to reduce duplicate work.

ratzdi replied:
done.

timon-ul replied:
You undid this change and need to redo it.

timon-ul commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -324,9 +313,9 @@ std::vector<SymbolTag> expandTagBitmask(const SymbolTags STGS) {
-  constexpr unsigned MinTag = static_cast<unsigned>(SymbolTag::FirstTag);
-  constexpr unsigned MaxTag = static_cast<unsigned>(SymbolTag::LastTag);
-  for (unsigned I = MinTag; I <= MaxTag; ++I) {
+  constexpr uint32_t MinTag = static_cast<uint32_t>(SymbolTag::FirstTag);
```
I know I pushed for this change, but I realized that `SymbolTag` does not inherit from `uint32_t` as I thought, but instead is an `int` as of now (yes, not `unsigned int`, and as far as I know this is implementation defined which type an enum class inherits from per default, so expecting any type is kinda a bad idea). So technically this change does not make sense really.  I am sorry. Also it kinda does not really matter what type you take here since it is to be expected that the enum will never fill too much for any type to fill. So `unsigned` is fine, but so is `int`.

ratzdi replied:
seems to be cleared?!

timon-ul replied:
I still see `uint32_t` everywhere.

timon-ul commented on clang-tools-extra/clangd/index/Symbol.h:
```diff
@@ -17,20 +17,108 @@
+///
+/// - The storage type adapts to SymbolTag's underlying type.
+/// - Bits outside the supported [FirstTag..LastTag] range are dropped.
+class SymbolTags {
```
I appreciate that you decided to go the extra mile to make this more type safe, but as of now I do not think it is doing it properly and just creates more confusion than safety. I might have been a bit naive when I was proposing this and I think doing this properly is a lot more nuanced.

The way you define the `Storage` type is not really doing anything for us.  `SymbolTag` is inheriting from `int` as of now and not representing its values as a bit mask. This means any information based on the type becomes near useless since it can contain as many values as an `int` can and creating storage depending on that is....pointless. We do not want a bit mask that can hold 4,294,967,295 values, that is way too much wasted space and I do not think `SymbolTag` will ever contain even a fraction of that many values. The only information that could be used to determine our type is what value `LastTag` has, and I at least do not see any nice way to make this type dependant on that. So I start to think that maybe returning to the old definition of `SymbolTags` is actually the simplest and easiest to understand.

So unless you come up with a better way of dealing with this problem I would suggest you either completly revert to the old definition for `SymbolTags` or you keep this class, but then there is quite some work to be done and here are my thoughts on them:

- why are the functions `fromRaw` and `fromTag` not constructors? Not that I have a particular strong opinion for or against this, I am just curious for your reasoning
- I do not like that you are "fixing" bad user input with `validMask()`. You should just assert for it being valid. Like if a user misuses this type and throws random stuff into it, then what even is the expected behaviour of you? We just pick the bits that are valid and pray that this is the way the user wants it to work? I think we should just clearly fail on those inputs and tell the user they are giving us garbage.
- I think we should only check for broken input if constructing from raw, everything else there should only be valid values unless someone does stuff like `(SymbolTag) 42` which yields a valid value (thank you C++), but I think dealing with that is too much.
- I think your `validMask` is broken, I have not 100% confirmmed this but I think you currently do not allow the first tag to be ever set.
- I do not like the option to implicitly cast to bool and would prefer it to be explicit.

ratzdi replied:
Without going into a detailed analysis of this comment, I agree with you, I would prefer the simplicity too.

ratzdi replied:
change is obsolete, due to revert.

ratzdi committed:
Replaced toSymbolTagBitmask with SymbolTags::fromTag.

ratzdi committed:
Simplified filterSymbolTags:

- Reused a combined mask (`RemoveVirtualAndOverrides`) to avoid repeating `~(Overrides | Virtual)`.
- Merged the two identical "remove virtual" cases (`Overrides`, `Abstract`) into one condition.

ratzdi committed:
Revert "Replaced toSymbolTagBitmask with SymbolTags::fromTag."

This reverts commit bdacf738a50a5ebd8c7ec69e14c8846ecfc3fae0.

ratzdi committed:
Revert "Simplified filterSymbolTags:"

This reverts commit 86a41ca30a044130d0cf774f3b3badf881f8c68a.

timon-ul commented:
@ratzdi you requested a review but you did not undo the changes we talked about? `SymbolTags` is still a class. Unless I misunderstood you.

ratzdi committed:
Revert "Refactor: Make SymbolTags type-safe and adapt tag handling"

This reverts commit 765ef3e7

ratzdi committed:
Reapply "Simplified filterSymbolTags:"

This reverts commit c39d64b9d01a8ac3504d1f8ce403336a2a5696ca.

ratzdi committed:
changed typecast from uint32_t to unsigned.

timon-ul commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -153,10 +192,55 @@ bool isUniqueDefinition(const NamedDecl *Decl) {
 
 SymbolTags toSymbolTagBitmask(const SymbolTag ST) {
-  return (1 << static_cast<unsigned>(ST));
+  return (1 << static_cast<uint32_t>(ST));
```
For consistency purposes you should revert this too.

ratzdi committed:
changed typecast from uint32_t to unsigned.

timon-ul reviewed (approved):
Looks good to me now, @HighCommander4 please have a look next :)

ArcsinX commented on clang-tools-extra/clangd/unittests/SerializationTests.cpp:
```diff
@@ -160,6 +161,10 @@ TEST(SerializationTest, YAMLConversions) {
   EXPECT_FALSE(Sym1.Flags & Symbol::Deprecated);
+  // Tags: Deprecated (1<<1=2) | Static (1<<8=256) = 258
+  EXPECT_EQ(Sym1.Tags, 258u);
+  EXPECT_TRUE(Sym1.Tags & (1u << static_cast<unsigned>(SymbolTag::Deprecated)));
```
`static_cast<unsigned>(SymbolTag::...)` appears quite frequently in this PR, but we have `toSymbolTagBitmask()` function for this. Maybe it's better to use `toSymbolTagBitmask()`?

ratzdi replied:
done.

ArcsinX commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -135,6 +138,42 @@ bool isFinal(const Decl *D) {
+// virtual)
+bool isOverrides(const NamedDecl *ND) {
+  if (const auto *MD = llvm::dyn_cast<CXXMethodDecl>(ND)) {
+    if (MD->size_overridden_methods() == 0)
```
I think we don't need to check `MD->size_overridden_methods()` here, because at line 150 we have a loop which will be just no-op, if there are not overridden methods.

ratzdi replied:
done.

ArcsinX commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -135,6 +138,42 @@ bool isFinal(const Decl *D) {
+
+    for (const auto *Overridden : MD->overridden_methods()) {
+      // Check if the overridden method is virtual but not pure virtual
+      if (Overridden->isVirtual() && !Overridden->isPureVirtual())
```
Instead of checking `Overridden->isVirtual()` at every loop iteration, we can check `MD->isVirtual()` before the loop and return false is not.
I mean that `Overridden` is virtual iff `MD` is virtual. Maybe inside this loop we can add an assertion for `Overridden->isVirtual()`.
I.e.
- check `MD->isVirtual()` before the loop
- add assertion for `Overridden->isVirtual()` inside the loop

ratzdi replied:
what do you mean with adding an assertion?

ArcsinX replied:
I mean that if `MD->isVirtual()` is TRUE, then `Overridden->isVirtual()` must be TRUE. That's why we don't need to check `Overridden->isVirtual()` inside the loop.
But we can add assert inside the loop for a case when we face a bug in the clang.
I.e.
```cpp
  if (const auto *MD = llvm::dyn_cast<CXXMethodDecl>(ND)) {
    if (!MD->isVirtual())
      return false;

    for (const auto *Overridden : MD->overridden_methods()) {
      assert(Overridden->isVirtual());
      if (!Overridden->isPureVirtual()))
      ............
    }
```

ratzdi replied:
done.

ArcsinX commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -135,6 +138,42 @@ bool isFinal(const Decl *D) {
+// 3. ALL overridden methods are pure virtual
+bool isImplements(const NamedDecl *ND) {
+  if (const auto *MD = llvm::dyn_cast<CXXMethodDecl>(ND)) {
+    if (MD->size_overridden_methods() == 0 || MD->isPureVirtual())
```
Same here. I think we don't need to check `MD->size_overridden_methods()`

ratzdi replied:
We have to, due to the specification 'It overrides at least one method'. If size_overridden_methods is zero FALSE must be returned. Not considering size_overridden_methods before the loop leads to wrong calculation, i.e. it returns true in that case.

ArcsinX replied:
Got it, thanks for clarification

ArcsinX commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -153,6 +192,51 @@ bool isUniqueDefinition(const NamedDecl *Decl) {
+  const SymbolTags AbstractMask = toSymbolTagBitmask(SymbolTag::Abstract);
+  const SymbolTags FinalMask = toSymbolTagBitmask(SymbolTag::Final);
+
+  const SymbolTags RemoveVirtualAndOverrides = VirtualMask | OverridesMask;
```
I think the name of this variable should be `VirtualAndOverridesMask`, reflecting what it holds.

ratzdi replied:
done.

ratzdi committed:
Refactor: Update symbol tag handling to use toSymbolTagBitmask for clarity. Improve isOverrides.

ratzdi committed:
fix: replaced Overridden->isVirtual() with assertion due to the severity of a clang-bug.

ArcsinX reviewed (approved):
LGTM
